### ISSUE TITLE
Re-implement as MTL

### ISSUE THREAD

**Issue by pbrisbin** [freckle/faktory_worker_haskell#60]
**Idea**: Re-implement internals in MTL-style

Optionally, continue to expose all the same `IO` versions.

## Benefits

Generally speaking, things that use `Monad(Unlift)IO` compose better into our
various applications. Using `MonadLogger` will also get Faktory-related logging
into the general stream with everything else.

Concretely, getting `perform` into a `MonadFaktoryPerform` type-class means that
enqueuing normally (`FaktoryPerformT`) or as a part of a batch (`FaktoryBatchT`)
will look exactly the same at the call-site (`perform`). This will also be a much
cleaner way to handle default `JobOptions`, which will prevent bugs.

Since we'll be implementing a new worker loop function, we can take this chance
to have it loop over `Job args` (not `args`), which is required if they want to
do things that require knowing Job metadata, such as submit progress updates
visible in `TRACK` calls or produce new Jobs into their own `BATCH`.

## Proposed Ergonomics

Given:

```hs
data Payload = Payload
  deriving stock Generic
  deriving anyclass (FromJSON, ToJSON)
```

### Producer / Perform

Current usage:

```hs
import Faktory.Job
import Faktory.Producer

main :: IO ()
main = bracket newProducerEnv closeProducer $ \producer ->
  perform empty producer Payload
```

Simplest MTL-ified usage:

```hs
import Control.Monad.Faktory
import Control.Monad.Faktory.Perform
import Control.Monad.Faktory.PerformDefault

main :: IO ()
main =
  runStderrLoggingT
    $ runFaktoryEnvT
    $ runFaktoryPerformT
    $ perform Payload
```

Real-world MTL usage:

```hs
-- Same imports as above

data App = App
  { --
  , appFaktoryClient :: Faktory.Client
  }

instance HasFaktoryClient App where
  faktoryClientL = lens appFaktoryClient $ \x y -> x { appFaktoryClient = y }

newtype AppT m a = AppT (ReaderT App m a)
  deriving newtype (..., MonadReader App)
  deriving MonadFaktory via (ActualFaktory m)
  deriving MonadFaktoryPerform via (DefaultFaktoryPerformm)

main :: IO ()
main = do
  app <- loadApp
  runApp app $ perform Payload
```

Doing a `perform` with explicit options not show. I'm thinking `local` would work to accomplish that:

```hs
local (faktoryJobOptionsL <>~ retry 3)) $ perform Payload
```

### Batch

Current usage:

```hs
import Faktory.Ent.Batch
import Faktory.Job
import Faktory.Producer

main :: IO ()
main = bracket newProducerEnv closeProducer $ \producer -> do
  onComplete <- buildJob mempty producer myJob

  runBatch (complete onComplete <> description "My Batch") producer $ do
    void $ batchPerform mempty producer Payload
    void $ batchPerform mempty producer Payload
```

Simplest MTL-ified usage:

```hs
import Control.Monad.Faktory
import Control.Monad.Faktory.Ent.Batch
import Control.Monad.Faktory.Perform
import Control.Monad.Faktory.PerformDefault

main :: IO ()
main =
  runStderrLoggingT $ runFaktoryEnvT $ runFaktoryPerformT $ do
    onComplete <- buildJob myJob

    runFaktoryBatchT (complete onComplete <> description "My Batch") $ do
      perform Payload
      perform Payload
```

Real-world MTL usage:

```hs
-- Same App setup as Perform
-- Same imports as above

main :: IO ()
main = do
  app <- loadApp
  runApp app $ do
    onComplete <- buildJob myJob

    runFaktoryBatchT (complete onComplete <> description "My Batch") $ do
      perform Payload
      perform Payload
```

### Worker

Current usage:

```hs
import Faktory.Worker

handle :: Payload -> IO ()
handle = undefined

main :: IO ()
main = runWorkerEnv $ handle
```

Simplest MTL-ified usage:

```hs
import Control.Monad.Faktory
import Control.Monad.Faktory.Worker

main :: IO ()
main =
  runStderrLoggingT
    $ runFaktoryEnvT
    $ runFaktoryWorkerEnvT
    $ liftIO . handle . jobArg
```

Real-world MTL usage:

```hs
-- Same App setup as Perform
-- Same imports as above

handle
  :: ( MonadIO m
     , MonadLogger m
     , MonadReader env m
     , MonadThis m
     , MonadThat m
     , HasThis env
     , HasThat env
     )
  => Job Payload
  -> m ()
handle = undefined

main :: IO ()
main = do
  app <- loadApp
  runApp app $ runFaktoryWorkerEnvT handle
```

## Rough Sketch

:warning: I got carried away writing the code, so expand at your own risk. What I
chose to flesh out may not be interesting, and I might've omitted important
stuff. **Continue at your own risk**.

**Control.Monad.Faktory**:

```hs
class Monad m => MonadFaktory m where
  commandByteString
    :: ByteString
    -> [ByteString]
    -> m (Either String (Maybe ByteString))

-- Implemented in terms of ^

command_ :: MonadFaktory m => ByteString -> [ByteString] -> m ()

commandOK :: MonadFaktory m => ByteString -> [ByteString] -> m ()

commandJSON
  :: FromJSON a => ByteString -> [ByteString] -> m (Either String (Maybe a))
```

**Control.Monad.Faktory.Actual**

Example instance, for use with `deriving via`:

```hs
class HasFaktoryClient env where
  faktoryClientL :: Lens' env Client

instance HasFaktoryClient Client where
  faktoryClientL = id

newtype ActualFaktory m a  = ActualFaktory
  { unActualFaktory :: m a
  }
  deriving newtype (..., MonadReader env)

instance (MonadIO m, MonadLogger m, MonadReader env m, HasFaktoryClient env)
    => MonadGitHub (ActualFaktory m) where
    commandByteString = do
      client <- view faktoryClientL
      liftIO $ -- the thing

newtype FaktoryT m a = FaktoryT (ReaderT Client m a)
  deriving MonadFaktory via (ActualFaktory (Faktory m)

runFaktoryT :: FaktoryT m a -> Client -> m a
runFaktoryT (FaktoryT r) = runReaderT r

runFaktoryEnvT :: MonadUnliftIO m => FaktoryT m a -> Client -> m a
runFaktoryEnvT = bracket newClientEnv closeClient runFaktoryT
```

**Control.Monad.Faktory.Perform**

```hs
class Monad m => MonadFaktoryPerform where
  defaultJobOptions :: m JobOptions
  defaultJobOptions = pure mempty

  perform :: Job args -> m ()
```

**Control.Monad.Faktory.Perform.Default**

```hs
newtype DefaultFaktoryPerform m a = DefaultFaktoryPerform
  { unDefaultFaktoryPerform :: m a
  }
  deriving newtype (..., MonadFaktory)

instance MonadFaktoryPerform (DefaultFaktoryPerform m) where
  perform args = do
    job <- buildJob defaultJobOptions args
    commandOK "PUSH" [encode job]

newtype FaktoryPerformT m a = FaktoryPerformT
  deriving MonadFaktoryPerform via (DefaultFaktoryPerform (FaktoryPerformT m))

runFaktoryPerformT :: FaktoryPerformT m a -> m a
runFaktoryPerformT (FaktoryPerformT r) = runReaderT r mempty
```

**Control.Monad.Faktory.Perform.Reader**

```hs
class HasFaktoryJobOptions env where
  faktoryJobOptionsL :: Lens' env JobOptions

instance HasFaktoryJobOptions JobOptions where
  faktoryJobOptionsL = id

newtype ReaderFaktoryPerform m a = ReaderFaktoryPerform
  { unReaderFaktoryPerform :: m a
  }
  deriving newtype (..., MonadFaktory, MonadReader env)

instance (MonadReader env m, HasFaktoryJobOptions env)
  => MonadFaktoryPerform (ReaderFaktoryPerform m) where
  defaultJobOptions = ask

  perform args = do
    job <- buildJob args -- uses defaultOptions
    commandOK "PUSH" [encode job]

newtype FaktoryPerformT m a = FaktoryPerformT (ReaderT JobOptions m a)
  deriving newtype (MonadReader JobOptions)
  deriving MonadFaktoryPerform via (ReaderFaktoryPerform (FaktoryPerformT m))

runFaktoryPerformT :: FaktoryPerformT m a -> JobOptions -> m a
runFaktoryPerformT (FaktoryPerformT r) = runReaderT r
```

**Control.Monad.Faktory.Batch**

Users would always interacts with concrete stack for this one.

```hs
newtype FaktoryBatchT m a = FaktoryBatchT (ReaderT BatchId m a)
  deriving newtype
    ( ...
    , MonadReader BatchId
    , MonadFaktory
    )

runFaktoryBatchT :: MonadFaktory m => FaktoryBatchT m a -> m a
runFaktoryBatchT (FaktoryBatchT r) = do
  batchId <- ...
  runReaderT r batchId

instance MonadFaktoryPerform m => MonadFaktoryPerform (FaktoryBatchT m) where
  defaultJobOptions = do
    batchId <- ask
    options <- lift defaultOptions
    pure $ options <> custom (CustomBatchId batchId)

  perform = lift . perform
```

**Control.Monad.Faktory.Worker**

```hs
data WorkerEnv -- WorkerId, WorkerSettings

newtype FaktoryWorkerT m a = FaktoryWorkerT
  { unFaktoryWorkerT :: ReaderT WorkerEnv m a
  }
  deriving newtype
    ( ...
    , MonadFaktory
    , MonadFaktoryPerform -- Just in case
    )

runFaktoryWorkerT
  :: (MonadUnliftIO m, MonadFaktory m)
  => (Job args -> FaktoryWorkerT m a)
  -> WorkerSettings
  -> m a
runFaktoryWorkerT f settings = do
  bracket (initializeWorker workerId) cleanupWorker $ \env ->
    flip runReaderT env $ unFaktoryWorkerT $ untilHalt $ do
      -- Thanks to MonadFaktory superclass
      emJob <- commandJSON "FETCH" [queueArg workerQueue]

      case emJob of
        -- Check for errors, handle delay loop, run f on the job call. ACK or FAIL

runFaktoryWorkerEnvT
  :: (MonadUnliftIO m, MonadFaktory m)
  => (Job args -> FaktoryWorkerT m a)
  -> m a
runFaktoryWorkerEnvT f = do
  settings <- workerSettingsEnv
  runFaktoryWorkerT f settings
```

**Comment by pbrisbin** [comment]
@eborden @cdparks I started down this track (I have the `FaktoryWorkerT` stuff going locally) as a way to implement `TRACK` with progress and make it easier to move things to `BATCH`. Realized it would be a huge change to see all the way though, so I decided to sketch it out completely first and get feedback. Thoughts?

**Comment by eborden** [comment]
Exposing a library via `mtl` has its tradeoffs (I mean exposing just `IO` also does). There is a lot of discussion in the community about what is best for a library. Exposing `IO` allows consumers to define their `mtl`, `free`, etc. abstraction over top to integrate with their machinery. Exposing `mtl` is more opinionated and carries other libraries with it. I've had plenty of friction with libraries that choose `MonadCatch`, `MonadBaseControl`, `MonadUnliftIO`, etc because that was not my choice. Exposing an opinionated `mtl` and an `IO` version is possibly a nice balance for a library.

---------------------------

In a more detailed realm, I'm concerned about this class.

```hs
class HasFaktoryClient env where
  faktoryClientL :: Lens' env Client
```

When I think about a `Client` and a web server, I expect that this resource will be used by many threads. As such I'm going to want it inside an `MVar` or a `Pool` to accommodate my concurrency needs. Maybe I want one, maybe I want many. My use case will define that. Having this bound to a single pure instance becomes problematic.

**Comment by pbrisbin** [comment]
> I've had plenty of friction with libraries that choose MonadCatch, MonadBaseControl, MonadUnliftIO, etc because that was not my choice

Same here, but part of the reason I find this to be a less controversial push today is that `MonadUnliftIO` has effectively won, IMO. Do you disagree?

> When I think about a Client and a web server, I expect that this resource will be used by many threads. As such I'm going to want it inside an MVar or a Pool to accommodate my concurrency needs. Maybe I want one, maybe I want many. My use case will define that. Having this bound to a single pure instance becomes problematic.

Well, we've (you've) already decided in the definition of `Client` itself. It hides an `MVar` and mutexes access.

```hs
data Client = Client
  { clientConnection :: MVar Connection
  , clientSettings :: Settings
  }
```

> Maybe I want one, maybe I want many. My use case will define that. Having this bound to a single pure instance becomes problematic.

Nothing here says you have to use `HasFaktoryClient` and derive `via` the `ActualFaktory` instance that relies on it. That is just an example / option for the simpler case. I'd argue that `MonadFaktory(commandByteString)` is offering exactly the flexible you describe here, you could hold onto as many clients as you want, mutex or not, and do whatever you like to issue a command.

**Comment by pbrisbin** [comment]
> Exposing an opinionated mtl and an IO version is possibly a nice balance for a library.

That's been easy enough to do so far, so I can consider it a requirement if you like.

**Comment by eborden** [comment]
> Well, we've (you've) already decided in the definition of Client itself. It hides an MVar and mutexes access.

Oh, turns out old me was also thinking about concurrency :laughing:.

> Nothing here says you have to use HasFaktoryClient and derive via the ActualFaktory instance that relies on it. That is just an example / option for the simpler case. I'd argue that MonadFaktory(commandByteString) is offering exactly the flexibility you describe here, you could hold onto as many clients as you want, mutex or not, and do whatever you like to issue a command.

I'm probably clinging too closely to an example. Though what I'm getting at is that I don't see us using `ActualFaktory` in our API server it only comes with a single `Client` instance. I really like the `DerivingVia` machinery for making deriving easy, so it might be useful to expose a `SingClientFaktory` and `ClientPoolFaktory` version.

I do agree that `MonadFaktory` is flexible enough for either of these circumstances.

-------------------------------------------

Is `MonadFaktoryPerform` only necessary because of `newJob` and `applyOptions`? I wonder if it is possible to fold those into `MonadFaktory`. Is it desired? The reason to split up methods into multiple classes is because they represent divergent paths.

Are these the case?
- If I'm a consumer I only need `MonadFaktory`.
- If I'm a producer I need both `MonadFaktory` and `MonadFaktoryPerform`.

**Comment by pbrisbin** [comment]
> Though what I'm getting at is that I don't see us using ActualFaktory in our API server if it only comes with a single Client instance.

It's still a good example, but not only would we do that, we do. This design is just refining from our actual usage.

> so it might be useful to expose a SingClientFaktory and ClientPoolFaktory version

Totally, if we find ourselves implementing that I would push to do it as another instance in this library.

> Is `MonadFaktoryPerform`...

Effectively, yes. Having default options used at `perform` time is useful (this comes from a direct need in our codebase), so that's what `Perform` is specifying. If we got rid of that, then

```hs
perform :: MonadFaktory m => JobOptions -> Job args -> m ()
```

is entirely possible. **However**, you lose the ability to move things to batching with this kind of diff::

```diff
-myOuterCode = do
+myOuterCode = runFaktoryBatch batchOptions $ do
   myUnawareCodeFarAway

 myUnawareCodeFarAway = perform Payload
```

In other words, we have to keep `batchPerform` vs `perform`, which another real pain point from our codebase that motivated me getting into all this.

> If I'm a consumer I only need MonadFaktory.

As long as you *only* consume. One can (and we do) produce things from within consuming loops, which would bring in the new class constraint.

> If I'm a producer I need both MonadFaktory and MonadFaktoryPerform.

Yes, which motivated `DefaultFaktoryProducer` as a simple derivable without any other requirements.

**Comment by eborden** [comment]
> not only would we do that, we do. This design is just refining from our actual usage.

This isn't actually true. We are spinning up a new `Client` for every single `enqueue` call.

```hs
enqueueJobs
  :: (MonadReader app m, HasFaktorySettings app, MonadUnliftIO m)
  => JobOptions
  -> NonEmpty Job
  -> m ()
enqueueJobs options jobs = Faktory.runProducer
  $ \client -> runConduit $ yieldMany jobs .| jobsSink options client
```

```hs
runProducer
  :: (MonadReader app m, HasFaktorySettings app, MonadUnliftIO m)
  => (Producer -> m a)
  -> m a
runProducer f = do
  settings <- asks getFaktorySettings
  runProducerWith settings f

-- | Run a Faktory action with the given settings
runProducerWith :: MonadUnliftIO m => Settings -> (Producer -> m b) -> m b
runProducerWith settings f =
  bracket (liftIO $ newProducer settings) (liftIO . closeProducer) f
```

We have a single data for the settings (`asks getFaktorySettings`), but we are unbound on connections. So far that hasn't bitten us, but a single `Client` could create lock contention as many producers attempt to enqueue at once and vying for ownership of the `MVar` inside the `Client`.

-----------------------------------

> In other words, we have to keep batchPerform vs perform, which another real pain point from our codebase that motivated me getting into all this.

Make sense! Thanks for helping me make the connection.

**Comment by eborden** [comment]
I'm remember some design stuff now. The `MVar` inside `Client` was put there to protect users from themselves. You can call many `command`s on a `Client`, but without the `MVar` you could have one `command` call listening to another `command` calls response. The internal `MVar` was never meant to serve as a pool, but as a safety measure as green threads interleave IO.

```hs
  sendUnsafe clientSettings conn cmd args
  -- Bad things could happen here without an `MVar`
  response <- recvUnsafe clientSettings conn
```

So it makes a `Client` thread safe, but it doesn't make it concurrency efficient.

**Comment by pbrisbin** [comment]
Thanks for clearing up the `Client` stuff. I could implement `ActualFaktory` to similarly create a client every time and require a `HasFaktorySettings` instance instead. But then the Worker implementation can't just re-use the ambient `MonadFaktory` since it needs to use the same client between fetch/ack/fail.

I have to think some more. It's possible this revelation doesn't change the core `MonadFaktory(Perform)` pattern but **does** mean the example instances should be different -- or maybe it means I should revisit entirely. Will come back next week.

**Comment by eborden** [comment]
I don't think it meaningfully changes `MonadFaktory` I think it is just a tweak to the examples.

**Comment by pbrisbin** [comment]
After thinking about this more, I'm actually going to call it *not* a good idea.

I think that this library should be lower-level operations that are pretty direct translations of the wire protocol, and living in `IO` is fine for that. This is how the other languages' client libraries work and I think we should target that. I plan to actually shift, and propose we use our next breaking version to change some names and functions to more directly mimic the other client libraries abstractions.

Things like `MonadFaktory` can be implemented by users of this library. If it becomes worth centralizing, we could make a separate `faktory-mtl` package (or something) to hold all that.

One thing I learned when investigating the other client libraries is that (at least) the Ruby library exposes a `Connection = Pool Client` object as the primary interface to call into the wire protocol. So maybe we should bring that into here instead.
